hy/hy_inspect.py

```diff
@@ -0,0 +1,238 @@
+"""
+Get useful information from live Hy or Python objects.
+
+This module provides Hy compatibility with CPython's inspect module.
+Within Hy, Hy macros are supported via Hy's `get_macro` macro,
+for example `(inspect.getsource (get-macro get-macro))`.
+The Hy-specific functionality is monkey-patched into `inspect` at run-time.
+
+The class finder in `findsource` relies on python 3.13 or later.
+"""
+
+import inspect
+import linecache
+import sys
+from contextlib import suppress
+
+from hy.compat import PY3_13
+from hy.errors import HySyntaxError
+from hy.models import as_model, Expression, Lazy, Object
+from hy.reader import HyReader, read
+from hy.reader.exceptions import LexException, PrematureEndOfInput
+
+
+class HySafeReader(HyReader):
+    """A HyReader that skips over non-default reader macros.
+
+    Skip over undefined reader macros so that no code is executed at read time.
+    Do this by replacing the tag reader method to return None, as comments do.
+    """
+
+    @reader_for(")")
+    @reader_for("]")
+    @reader_for("}")
+    def _skip(self, key):
+        return None
+
+    @reader_for("#")
+    def _safe_tag_dispatch(self, key):
+        """Skip over undefined reader macros so that no code is executed at read time."""
+        with suppress(LexException):
+            return super().tag_dispatch(key)
+
+
+def isExpression(object):
+    """Check if object is a Hy Expression instance."""
+    return isinstance(object, Expression)
+
+
+def isLazy(object):
+    """Check if object is a Hy Lazy instance."""
+    return isinstance(object, Lazy)
+
+
+def getfile(object):
+    """Work out which source or compiled file an object was defined in."""
+
+    if isLazy(object) or isExpression(object):
+        if hasattr(object, "filename"):
+            return object.filename
+        else:
+            # python's inspect.getfile raises OSError where there's no __file__
+            raise OSError("source code not available")
+
+    try:
+        return py_getfile(object)
+    except TypeError:
+        # python's inspect.getfile raises TypeError for unhandled types
+        raise TypeError(
+            "module, class, method, function, traceback, frame, "
+            "code, Expression or Lazy object was expected, got "
+            "{}".format(type(object).__name__)
+        )
+
+
+def findsource(object):
+    """Return the entire source file and starting line number for an object.
+
+    First looks for Hy source, otherwise defers to the original
+    `inspect.findsource`. The argument may be a module, class, method,
+    function, traceback, frame, code, Lazy or Expression object. The source
```

Comment:
> Lazy or Expression object

doesn't seem to be true; passing in the output of `hy.read` or `hy.read-all` throws `AttributeError: 'NoneType' object has no attribute 'endswith'` (from `inspect.getsourcefile` below).

---

in any case, is there a reason we need this to work on models? I'd expect `inspect` functions to only be called on (compiled) objects (@atisharma ?)